okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -1243,14 +1243,32 @@ private static String canonicalizeHost(String input, int pos, int limit) {
       if (idnDecoded == null) return null;
 
       // Confirm that the decoded result doesn't contain any illegal characters.
-      int length = idnDecoded.length();
-      if (delimiterOffset(idnDecoded, 0, length, "\u0000\t\n\r #%/:?@[\\]") != length) {
+      if (containsInvalidHostnameCharacters(idnDecoded)) {
         return null;
       }
 
       return idnDecoded;
     }
 
+    private static boolean containsInvalidHostnameCharacters(String idnDecoded) {
+      for (int i = 0; i < idnDecoded.length(); i++) {
```

Comment:
wanna use a codepoint loop instead of a char loop? There's a few other examples in this file.